README.md
# angular-ujs
> Distributed via

[![Version     ](https://img.shields.io/gem/v/angular-ujs.svg)                                      ](https://rubygems.org/gems/angular-ujs)

> Unobtrusive scripting for AngularJS ( without jQuery dependency )

[![Travis CI   ](https://travis-ci.org/tomchentw/angular-ujs.svg?branch=master)                     ](https://travis-ci.org/tomchentw/angular-ujs)
[![Quality     ](https://img.shields.io/codeclimate/github/tomchentw/angular-ujs.svg)               ](https://codeclimate.com/github/tomchentw/angular-ujs)
[![Coverage    ](https://img.shields.io/coveralls/tomchentw/angular-ujs.svg)                        ](https://coveralls.io/r/tomchentw/angular-ujs)
[![Dependencies](https://gemnasium.com/tomchentw/angular-ujs.svg)                                   ](https://gemnasium.com/tomchentw/angular-ujs)


## Project philosophy

### Develop in LiveScript
[LiveScript](http://livescript.net/) is a compile-to-js language, which provides us more robust way to write JavaScript.  
It also has great readibility and lots of syntax sugar just like you're writting python/ruby.

### Native, lightweight directives
Unobtrusive scripting in `jquery_ujs` provides the same interface with angular `directives`.  
We use the similarity between them and provides seamless intergration with `jquery_ujs`.  

### Spec / Scenario coverage
We use `krama` to run unit test against [angular-ujs.spec.ls](https://github.com/tomchentw/angular-ujs/blob/master/src/angular-ujs.spec.ls) and use `protractor` to run intergration test via [angular-ujs.scenario.ls](https://github.com/tomchentw/angular-ujs/blob/master/src/angular-ujs.scenario.ls).


## Installation

We follow **DRY** and has one dependency only on `angularjs`.  
However, we recommend you add [`ng-rails-csrf`](https://github.com/xrd/ng-rails-csrf/) into your project. As it name suggests, `ng-rails-csrf` automatically resolves CSRF in `angularjs` environment without `jquery_ujs`.

### Just use it

* (_Optional_) Download and include [`ng-rails-csrf.js`](https://github.com/xrd/ng-rails-csrf/blob/master/vendor/assets/javascripts/ng-rails-csrf.js).
* Download and include [`angular-ujs.js`](https://github.com/tomchentw/angular-ujs/blob/master/angular-ujs.js) OR [`angular-ujs.min.js`](https://github.com/tomchentw/angular-ujs/blob/master/angular-ujs.min.js).  

Then include them through script tag in your HTML.

### **Rails** projects (Only support 3.1+)
Add this line to your application's Gemfile:

```ruby
gem 'ng-rails-csrf' # Optional
gem 'angular-ujs'
```

And then execute:

    $ bundle

Then add these lines to the top of your `app/assets/javascripts/application.js` file:

```javascript
//= require angular
//= require ng-rails-csrf (Optional)
//= require angular-ujs
```

And include in your `angular` module definition:

```javascript
/* 'angular.ujs' DO NOT depend on 'ng-rails-csrf' module.
* You need to include it yourself.
*/  
var module = angular.module('my-awesome-project', ['angular.ujs']).
```


## Usage

### "data-confirm": Confirmation dialogs for links and forms

```html
<form data-confirm="Are you sure you want to submit?">...</form>
```

### "data-method": Links that result in POST, PUT, or DELETE requests

```html
<a href="..." data-method="delete" rel="nofollow">Delete this entry</a>
```

### "data-remote": Make links and forms submit asynchronously with Ajax
** Notice : API changed **

For `angularjs` apps, **ONLY** those items with `ng-model` will be submitted with `data-remote`

```html
<form data-remote="true" action="...">
  <input type="text" name="name" ng-model="name">
</form>
```

#### You can specify the model name via `data-remote` :
```html
<form data-remote="user" action="...">
  <input type="text" name="name" ng-model="user.name">
  <input type="email" name="email" ng-model="user.email">
</form>
```

### Use them all together :

```html
<a href="..." data-method="delete" data-remote="true" data-confirm="Are you sure you want to delete?" rel="nofollow">Delete this entry</a>
```


## Contributing

[![devDependency Status](https://david-dm.org/tomchentw/angular-ujs/dev-status.svg?theme=shields.io)](https://david-dm.org/tomchentw/angular-ujs#info=devDependencies)

1. Fork it ( https://github.com/tomchentw/angular-ujs/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

CHANGELOG.md
<a name="0.4.13"></a>
### 0.4.13  (2014-04-23)


#### Bug Fixes

* **Makefile:** check TRAVIS defined ([3659f970](https://github.com/tomchentw/angular-ujs/commit/3659f970571635a16541763eab16699edda1a338))


<a name="0.4.12"></a>
### 0.4.12 (2014-03-14)


<a name="0.4.11"></a>
### 0.4.11 (2014-03-14)


#### Bug Fixes

* **Makefile:** start test/scenario-rails ([3fe09534](https://github.com/tomchentw/angular-ujs/commit/3fe09534f4b68689ff8aece6ef7f21120f801ec4))
* **README.md:** update coveralls link ([ae07b0a7](https://github.com/tomchentw/angular-ujs/commit/ae07b0a70737e5a955180f026a865bde58503637))


<a name="0.4.10"></a>
### 0.4.10 (2014-03-09)


#### Bug Fixes

* **travis:** typo ([b3bbef1f](https://github.com/tomchentw/angular-ujs/commit/b3bbef1f85c830eedeb9a4c623ff6e8809d48610))


<a name="0.4.9"></a>
### 0.4.9 (2014-02-24)


<a name="0.4.8"></a>
### 0.4.8 (2014-02-11)


#### Bug Fixes

* **bower.json:** update dependencies and devDependencies ([0912c85aa8c962f76f30756c002f6acb490f9304](https://github.com/tomchentw/angular-ujs/commit/0912c85aa8c962f76f30756c002f6acb490f9304))


#### Features

* **gulpfile:** adjust for gulp 3.5.0 from tomchentw-boilerplate ([b5d7a69145dc51a29c3cfbb474b4778a8597c21e](https://github.com/tomchentw/angular-ujs/commit/b5d7a69145dc51a29c3cfbb474b4778a8597c21e))


<a name="0.4.7"></a>
### 0.4.7 (2014-01-25)


#### Features

* **spec:** add test coverages ([35b186acaed43bd5c4e910ecdace485fe2460ff8](https://github.com/tomchentw/angular-ujs/commit/35b186acaed43bd5c4e910ecdace485fe2460ff8))


<a name="0.4.6"></a>
### 0.4.6 (2014-01-17)


#### Bug Fixes

* **gulpfile:** add release-git task with `git push`, `--tags` ([3d965954f04da8826d90dc86755d96c6247412fa](https://github.com/tomchentw/angular-ujs/commit/3d965954f04da8826d90dc86755d96c6247412fa))
* **travis:** start rails server in before_script ([c7c178336375419339867fe9cd01ac43e4700d3a](https://github.com/tomchentw/angular-ujs/commit/c7c178336375419339867fe9cd01ac43e4700d3a))


<a name="0.4.5"></a>
### 0.4.5 (2014-01-16)


<a name="0.4.4"></a>
### 0.4.4 (2014-01-13)


#### Bug Fixes

* **travis:** start rails server in before_script ([c7c178336375419339867fe9cd01ac43e4700d3a](https://github.com/tomchentw/angular-ujs/commit/c7c178336375419339867fe9cd01ac43e4700d3a))


<a name="v0.4.3"></a>
### v0.4.3 (2014-01-04)


#### Bug Fixes

* **CHANGELOG:** should be v0.4.2 ([bf33e9ce](http://github.com/tomchentw/angular-ujs/commit/bf33e9ce1ac1b34ffb2661d96af0c757819fb4a2))

<a name="v0.4.2"></a>
### v0.4.2 (2014-01-04)


#### Bug Fixes

* **confirm:** deny default action only when dismiss confirm ([bcbdd425](http://github.com/tomchentw/angular-ujs/commit/bcbdd42552ea2850693510c48c7df5f9c915b19a))


<a name="0.3.x"></a>
### 0.3.x (2013-12-31)

The changelog before `v0.4.0` are collapsed into one.

#### Bug Fixes

* **Gruntfile:** add shell prefix ([73f2381c3316a826d33720f0d00be26a70ae5506](https://github.com/tomchentw/angular-ujs/commit/73f2381c3316a826d33720f0d00be26a70ae5506))
* **confirm:**
  * directive should inject RailsConfirmCtrl ([bfe2571afbda23189fa10a59ab7032a75774c22f](https://github.com/tomchentw/angular-ujs/commit/bfe2571afbda23189fa10a59ab7032a75774c22f))
  * fix confirm logic and add confirm directive ([903edd646c41d4faf639d9f790516e02ca46febd](https://github.com/tomchentw/angular-ujs/commit/903edd646c41d4faf639d9f790516e02ca46febd))
* **directive:** fix compile return noop, change to void ([3d31a30474a683f436f88e0df6032d7afa55819a](https://github.com/tomchentw/angular-ujs/commit/3d31a30474a683f436f88e0df6032d7afa55819a))
* **rails:**
  * createMethodFormElement will use $apply ([77d8c416bdb3a33a071c075147352a84cb74ff3c](https://github.com/tomchentw/angular-ujs/commit/77d8c416bdb3a33a071c075147352a84cb74ff3c))
  * remove jQuery method ([1c2f004c43321a97eca5f0381df1a83d67d21608](https://github.com/tomchentw/angular-ujs/commit/1c2f004c43321a97eca5f0381df1a83d67d21608))
* **remote:** rails 4 http method overriding issue and make them work together ([ed00160862635e985b1b56ce798ee4e402994e0b](https://github.com/tomchentw/angular-ujs/commit/ed00160862635e985b1b56ce798ee4e402994e0b))
* **travis:** run db:migrate before start server ([c19387daaea1e101bb4a640fe533dcd13d648e86](https://github.com/tomchentw/angular-ujs/commit/c19387daaea1e101bb4a640fe533dcd13d648e86))


#### Features

* **$getRailsCSRF:** add conditional service ([bc39820863b63e93eff1f200ef44c70d8c08ca00](https://github.com/tomchentw/angular-ujs/commit/bc39820863b63e93eff1f200ef44c70d8c08ca00))
* **controller:** add RailsRemoteFormCtrl ([fa71046595028dc39a21f0df9e96caa3354bb137](https://github.com/tomchentw/angular-ujs/commit/fa71046595028dc39a21f0df9e96caa3354bb137))
* **directive:**
  * remote directive will evaluation on $scope if data-remote is true ([041c8f64a245651a002554321c236a31a28c9374](https://github.com/tomchentw/angular-ujs/commit/041c8f64a245651a002554321c236a31a28c9374))
  * add method directive and work with remote directive ([fa05cc5889e88a9391d2769c7bd1f4462d6f6d4c](https://github.com/tomchentw/angular-ujs/commit/fa05cc5889e88a9391d2769c7bd1f4462d6f6d4c))
  * add remote directive ([1e5166679f81db4b0e3442a3c1eb035cb0cb5b85](https://github.com/tomchentw/angular-ujs/commit/1e5166679f81db4b0e3442a3c1eb035cb0cb5b85))
* **karma:** add karma and jasmine as test framework, remove .min.js in vender ([73c1601ba45f8aca2933f6c39b8ff7340778f880](https://github.com/tomchentw/angular-ujs/commit/73c1601ba45f8aca2933f6c39b8ff7340778f880))
* **noopRailsConfirmCtrl:** extract to controller ([a4ec390443e3eee0c463f198a65422a3b4c4befb](https://github.com/tomchentw/angular-ujs/commit/a4ec390443e3eee0c463f198a65422a3b4c4befb))
* **noopRailsRemoteFormCtrl:** extract to controller ([90189f77e2c57a57fd2d26aa844f8dc32cd08d24](https://github.com/tomchentw/angular-ujs/commit/90189f77e2c57a57fd2d26aa844f8dc32cd08d24))
* **rails:**

angular-ujs.js
/*! angular-ujs - v 0.4.13 - Wed Apr 23 2014 21:24:18 GMT+0800 (CST)
 * https://github.com/tomchentw/angular-ujs
 * Copyright (c) 2014 [tomchentw](https://github.com/tomchentw);
 * Licensed [MIT](http://tomchentw.mit-license.org)
 */
/*global angular:false*/
(function(angular, bind){
  'use strict';
  function denyDefaultAction(event){
    event.preventDefault();
    event.stopPropagation();
  }
  angular.module('angular.ujs', []).factory('$getRailsCSRF', ['$document'].concat(function($document){
    return function(){
      var metas, i$, ref$, len$, meta;
      metas = {};
      for (i$ = 0, len$ = (ref$ = $document[0].querySelectorAll('meta[name^="csrf-"]')).length; i$ < len$; ++i$) {
        meta = ref$[i$];
        meta = angular.element(meta);
        metas[meta.attr('name')] = meta.attr('content');
      }
      return metas;
    };
  })).controller('noopRailsConfirmCtrl', (function(){
    var prototype = constructor.prototype;
    prototype.allowAction = function(){
      return true;
    };
    prototype.denyDefaultAction = denyDefaultAction;
    function constructor(){}
    return constructor;
  }())).controller('RailsConfirmCtrl', (function(){
    var prototype = constructor.prototype;
    prototype.allowAction = function(){
      var message;
      message = this.$attrs.confirm;
      return angular.isDefined(message) && this.$window.confirm(message);
    };
    prototype.denyDefaultAction = denyDefaultAction;
    constructor.$inject = ['$window', '$attrs'];
    function constructor($window, $attrs){
      this.$window = $window;
      this.$attrs = $attrs;
    }
    return constructor;
  }())).controller('noopRailsRemoteFormCtrl', (function(){
    var prototype = constructor.prototype;
    prototype.submit = function($form){
      $form[0].submit();
      return {
        then: angular.noop
      };
    };
    function constructor(){}
    return constructor;
  }())).controller('RailsRemoteFormCtrl', (function(){
    var prototype = constructor.prototype;
    prototype.submit = function($form){
      var targetScope, modelName, data, key, value, config, METHOD, own$ = {}.hasOwnProperty;
      targetScope = $form.scope();
      modelName = this.$attrs.remote;
      data = {};
      if (modelName + "" !== 'true') {
        this.$parse(modelName).assign(data, targetScope.$eval(modelName));
      } else {
        for (key in targetScope) if (own$.call(targetScope, key)) {
          value = targetScope[key];
          if (key === 'this' || key[0] === '$') {
            continue;
          }
          data[key] = value;
        }
      }
      config = {
        url: $form.attr('action'),
        method: $form.attr('method'),
        data: data
      };
      METHOD = data._method;
      if (METHOD !== 'GET' && METHOD !== 'POST') {
        config.headers = {
          'X-Http-Method-Override': METHOD
        };
      }
      return this.$http(config).then(this.successCallback, this.errorCallback);
    };
    prototype.successCallback = function(response){
      this.$scope.$emit('rails:remote:success', response);
    };
    prototype.errorCallback = function(response){
      this.$scope.$emit('rails:remote:error', response);
    };
    constructor.$inject = ['$scope', '$attrs', '$parse', '$http'];
    function constructor($scope, $attrs, $parse, $http){
      this.$scope = $scope;
      this.$attrs = $attrs;
      this.$parse = $parse;
      this.$http = $http;
      this.successCallback = bind(this, this.successCallback);
      this.errorCallback = bind(this, this.errorCallback);
    }
    return constructor;
  }())).directive('confirm', function(){
    function onClickHandler(confirmCtrl, event){
      if (!confirmCtrl.allowAction()) {
        confirmCtrl.denyDefaultAction(event);
      }
    }
    function postLinkFn($scope, $element, $attrs, $ctrls){
      var callback;
      callback = bind(void 8, onClickHandler, $ctrls[0]);
      $element.on('click', callback);
      $scope.$on('$destroy', function(){
        $element.off('click', callback);
      });
    }
    return {
      restrict: 'A',
      require: ['confirm'],
      controller: 'RailsConfirmCtrl',
      compile: function(tElement, tAttrs){
        var $attr;
        $attr = tAttrs.$attr;
        if ($attr.confirm !== 'data-confirm' || $attr.remote === 'data-remote' || $attr.method === 'data-method') {
          return;
        }
        return postLinkFn;
      }
    };
  }).directive('remote', ['$controller'].concat(function($controller){
    function onSubmitHandler($element, $ctrls, event){
      $ctrls[1].denyDefaultAction(event);
      if ($ctrls[1].allowAction()) {
        $ctrls[0].submit($element);
      }
    }
    function postLinkFn($scope, $element, $attrs, $ctrls){
      var callback;
      if (!$ctrls[1]) {
        $ctrls[1] = $controller('noopRailsConfirmCtrl', {
          $scope: $scope
        });
      }
      callback = bind(void 8, onSubmitHandler, $element, $ctrls);
      $element.on('submit', callback);
      $scope.$on('$destroy', function(){
        $element.off('submit', callback);
      });
    }
    return {
      require: ['remote', '?confirm'],
      restrict: 'A',
      controller: 'RailsRemoteFormCtrl',
      compile: function(tElement, tAttrs){
        if (tAttrs.$attr.remote !== 'data-remote') {
          return;
        }
        return postLinkFn;
      }
    };
  })).directive('method', ['$controller', '$compile', '$document', '$getRailsCSRF'].concat(function($controller, $compile, $document, $getRailsCSRF){
    function onClickHandler($scope, $attrs, $ctrls, event){
      var metaTags, childScope, $form;
      if ($ctrls[0].allowAction()) {
        $ctrls[0].denyDefaultAction(event);
      }
      metaTags = $getRailsCSRF();
      childScope = $scope.$new();
      $form = $compile("<form class=\"ng-hide\" method=\"POST\" action=\"" + $attrs.href + "\">\n  <input type=\"text\" name=\"_method\" ng-model=\"_method\">\n  <input type=\"text\" name=\"" + metaTags['csrf-param'] + "\" value=\"" + metaTags['csrf-token'] + "\">\n</form>")(childScope);
      $document.find('body').append($form);
      childScope.$apply(function(){
        childScope._method = $attrs.method;
      });
      $ctrls[1].submit($form).then(function(){
        childScope.$destroy();
        $form.remove();
      });
    }
    function postLinkFn($scope, $element, $attrs, $ctrls){
      var controllerArgs, callback;
      controllerArgs = {
        $scope: $scope,
        $attrs: $attrs
      };
      if (!$ctrls[0]) {
        $ctrls[0] = $controller('noopRailsConfirmCtrl', controllerArgs);
      }
      if (!$ctrls[1]) {
        $ctrls[1] = $controller('noopRailsRemoteFormCtrl', controllerArgs);
      }
      callback = bind($ctrls, onClickHandler, $scope, $attrs, $ctrls);
      $element.on('click', callback);
      $scope.$on('$destroy', function(){
        $element.off('click', callback);
      });
    }
    return {
      require: ['?confirm', '?remote'],
      restrict: 'A',
      compile: function(tElement, tAttrs){
        if (tAttrs.$attr.method !== 'data-method') {
          return;
        }
        return postLinkFn;
      }
    };
  }));
}.call(this, angular, angular.bind));